### PR TITLE
Joseph Mock

### PR DESCRIPTION
**Pull request by JoeyMock**

### CODE REVIEW

File: brevets/acp_times.py [hunk 1]
```diff
@@ -26,7 +26,38 @@ def open_time(control_dist_km, brevet_dist_km, brevet_start_time):
...
+        return timeRet
+
+    if(brevet_dist_km > 400):
+        timeRet.shift(hours=400/34)
+        brevet_dist_km -= 400
+    else:
+        timeRet.shift(hours=brevet_dist_km/28)
+        return timeRet
+
+    return timeRet
```
Comment (thejanjan) [inline brevets/acp_times.py:60]: this algorithm is ok but you can optimize it with magic number tables as talked about a bit in class, by creating a dictionary that maps the brevet distances to their speeds & basing the algorithm off of that. also minor nitpick: avoid mixing snake/camel case for variable names

File: brevets/acp_times.py [hunk 2]
```diff
@@ -41,4 +72,35 @@ def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
...
+        return timeRet
+
+    if(brevet_dist_km > 400):
+        timeRet.shift(hours=400/11.428)
+        brevet_dist_km -= 400
+    else:
+        timeRet.shift(hours=brevet_dist_km/11.428)
+        return timeRet
+
+    return timeRet
```
Comment (thejanjan) [inline brevets/acp_times.py:106]: same as above

File: brevets/tests/test_acp_times.py [hunk 3]
```diff
@@ -6,6 +6,26 @@
...
+    assert open_time(205, 200, arrow.now()) == arrow.now().shift(hours = 200/34)
+
+def test_less():
+    assert open_time(190, 200, arrow.now()) == arrow.now().shift(hours = 190/34)
+
+def test_bigger():
+    assert open_time(300, 300, arrow.now()) == arrow.now().shift(hours = (200/34+100/32))
+
+def test_eBigger():
+    assert open_time(600, 600, arrow.now()) == arrow.now().shift(hours = (200/34+200/32+200/30))
```
Comment (thejanjan) [inline brevets/tests/test_acp_times.py:31]: not enough close time cases imo, and also some more descriptive test names would be beneficial

**Review by thejanjan — APPROVED** [review]

File: brevets/acp_times.py [hunk 4]
```diff
@@ -26,7 +26,38 @@ def open_time(control_dist_km, brevet_dist_km, brevet_start_time):
        An arrow object indicating the control open time.
        This will be in the same time zone as the brevet start time.
     """
-    return arrow.now()
+    timeRet = brevet_start_time
+    if(control_dist_km > brevet_dist_km):
+        control_dist_km = brevet_dist_km
+    if(brevet_dist_km > 200):
```
Comment (55wans05) [inline brevets/acp_times.py:32]: I concur with the comment below. Using a table would probably be more efficient. But there's another benefit. If you needed to modify the algorithm later, or have to implement variations in calculations, then using a table-based structure would probably make your life easier in the future.

File: brevets/acp_times.py [hunk 5]
```diff
@@ -41,4 +72,35 @@ def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
...
+        brevet_dist_km -= 200
+    else:
+        timeRet.shift(hours=brevet_dist_km/15)
+        return timeRet
+
+    if(brevet_dist_km > 400):
+        timeRet.shift(hours=400/11.428)
+        brevet_dist_km -= 400
+    else:
+        timeRet.shift(hours=brevet_dist_km/11.428)
```
Comment (55wans05) [inline brevets/acp_times.py:103]: One thing to keep in mind here, and in this project as a whole, is the danger of float arithmetic and how computers are just bad at doing float math. So it's probably worthwhile to have explicit rounding and such. Especially when you're doing unit testing so that your output is more predictable.

File: brevets/flask_brevets.py [hunk 6]
```diff
@@ -51,14 +52,18 @@ def _calc_times():
...
+    app.logger.debug("dist: {}".format(dist))
     app.logger.debug("km={}".format(km))
     app.logger.debug("request.args: {}".format(request.args))
     # FIXME!
     # Right now, only the current time is passed as the start time
     # and control distance is fixed to 200
     # You should get these from the webpage!
-    open_time = acp_times.open_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
-    close_time = acp_times.close_time(km, 200, arrow.now().isoformat).format('YYYY-MM-DDTHH:mm')
+    open_time = acp_times.open_time(km, dist, arrow.now()).format('YYYY-MM-DDTHH:mm')
```
Comment (55wans05) [inline brevets/flask_brevets.py:65]: Good work here! In the future perhaps adding input checking then responding to bad input with a 400 error would be helpful as the project gets more complicated.

File: brevets/tests/test_acp_times.py [hunk 7]
```diff
@@ -6,6 +6,23 @@
...
 import nose    # Testing framework
 import logging
+import ../acp_times.py
 logging.basicConfig(format='%(levelname)s:%(message)s',
                     level=logging.WARNING)
 log = logging.getLogger(__name__)
+
+def test first():
+    assert  open_time(200, 200, arrow.now()) == arrow.now().shift(hours = 200/34)
+    assert  close_time(200, 200, arrow.now()) == arrow.now().shift(hours = 200/15)
```
Comment (55wans05) [inline brevets/tests/test_acp_times.py:19]: Good work on tests here. I feel like you might want to consider the edge cases a little bit more and it seems like the close_time function might be a little undertested. But to be honest, for the purposes of this class, this is more than sufficient (:

**Review by 55wans05 — APPROVED** [review]
All told, good work! I think most of my comments are really future oriented because this project is a great baseline, and there weren't many obvious things that went wrong. So good job!
